ts/backend/src/app/features/utils/aggregator.mts

```diff
@@ -0,0 +1,500 @@
+
+/*
+REMARK:
+
+Definition tree is like this:
+
+BenchmarkGroupDefinition - a group of benchmark definitions
+└ BenchmarkDefinition[] - benchmark, group of test definitions
+  └ TestDefinition[] - available tests
+    └ ScenarioDefinition[] - defined scenarios per test
+
+
+Aggregated/scored structure is like this:
+
+BenchmarkGroupScored - a group of benchmark definitions
+└ BenchmarkScored[] - scored benchmarks
+  └ SubmissionScored[] - scored submissions
+    └ TestScored[] - scored tests
+      └ ScenarioScored[] - defined scenarios per test
+
+N.B. the scored tree knows a level between benchmarks and tests (submissions).
+
+Ideally, the two trees would match in overall structure. Currently this is
+achieved by moving the field definitions from BenchmarkDefinition to a
+separate `SubmissionDefinition` in `getBenchmarkScored`.
+
+(I would like to normalize that, i.e. introduce `SubmissionDefinition` in the
+DB schema.)
+
+Question: Could a benchmark possibly have multiple SubmissionDefinitions (then
+the submission would have to explicitly link to that) or is there always exactly
+one of those per benchmark (then the SubmissionDefinition can be inferred from
+a submission's linked benchmark).
+
+*/
+
+export interface FieldDefinition {
+  /** Name of field, how it's exposed */
+  out_field: string
+  // REMARK: introduced order to be able to define primary/secondary etc for
+  // sorting when a score is missing (this is hypothetical RN, haven't really
+  // thought through if that case is possible)
+  /**
+   * Order of field, 0 indicates the primary score.
+   * Optional, falls back to order of appearance.
+   */
+  order?: number
+  /**
+   * Aggregation function to use for aggregated fields (optional, if undefined
+   * the field score is taken from the table directly)
+   */
+  agg_func?: string
+  in_fields?: string | string[]
+  weights?: number[]
+  agg_lateral?: boolean
+}
+
+export interface ScenarioDefinition {
+  id: string
+  test_id: string
+  // REMARK: redundant:
+  // benchmark_id: string
+  name: string
+  description: string
+  fields: FieldDefinition[]
+}
+
+export interface TestDefinition {
+  id: string
+  benchmark_id: string
+  name: string
+  fields: FieldDefinition[]
+}
+
+// REMARK: not yet a thing - see first remark
+export interface SubmissionDefinition {
+  fields: FieldDefinition[]
+}
+
+export interface BenchmarkDefinition {
+  id: string
+  name: string
+  // REMARK: field definitions for aggregation on submission level
+  submission_fields: FieldDefinition[]
+  // REMARK: field definitions for aggregation on benchmark level
+  fields: FieldDefinition[]
+}
+
+export interface Result {
+  scenario_id: string
+  submission_id: string
+  key: string
+  score: number
+}
+
+export interface Submission {
+  id: string
+  status: 'SUBMITTED' | 'RUNNING' | 'SUCCESS' | 'FAILURE'
+  benchmark_id: string
+  description: string
+  published: boolean
+  tests: string[]
+}
+
+export interface Scoring {
```

Comment:
Is this a data structure for aggregation only? What's the corresponding persisted data type? There, we would need link to SubmissionId as well.